packages/hardhat-core/test/internal/hardhat-network/provider/node.ts

```diff
@@ -730,6 +737,144 @@ describe("HardhatNode", () => {
       });
     }
   });
+
+  it("should run calls in the right hardfork context", async function () {
+    // fork mainnet at the block when EIP-1559 activated (12965000), and try to
+    // run a call that specifies gas limits in EIP-1559 terms, but run that
+    // call one block earlier, and expect that call to fail because it should
+    // have specified gas limits in PRE-EIP-1559 terms.
+
+    this.timeout(10000);
+
+    // as a test that does forking, we need a remote Alchemy node to fork from:
+    if (ALCHEMY_URL === undefined) {
+      return;
+    }
+    const urlOfNodeToFork = ALCHEMY_URL;
+
+    const eip1559ActivationBlock = 12965000;
+
+    const forkedNodeConfig: ForkedNodeConfig = {
+      automine: true,
+      networkName: "mainnet",
+      chainId: 1,
+      networkId: 1,
+      hardfork: "london",
+      forkConfig: {
+        jsonRpcUrl: urlOfNodeToFork,
+        blockNumber: eip1559ActivationBlock,
+      },
+      forkCachePath: FORK_TESTS_CACHE_PATH,
+      blockGasLimit: 1_000_000,
+      minGasPrice: new BN(0),
+      genesisAccounts: [],
+      chains: defaultHardhatNetworkParams.chains,
+    };
+
+    const [, regularNode] = await HardhatNode.create(forkedNodeConfig);
+
+    const nodeCfgWithEarlyLondon = forkedNodeConfig;
+    let earlyLondonMainnetConfig = forkedNodeConfig.chains.get(1);
+    if (earlyLondonMainnetConfig === undefined) {
+      earlyLondonMainnetConfig = { hardforkHistory: new Map() };
+    }
+    earlyLondonMainnetConfig.hardforkHistory.set(
```

Comment:
Isn't this modifying (by shared references) the `defaultHardhatNetworkParams.chains` map?

Even if it's not, it isn't obvious to me that it isn't. Maybe this could start with something like:

```js
const nodeCfgWithEarlyLondon = {
  ...forkedNodeConfig,
  chains: new Map(forkedNodeConfig.chains)
}
```

just to be sure.

Reply:
Wait, no, that will do a shallow copy. Oh well.

Reply:
Addressed in https://github.com/nomiclabs/hardhat/pull/1928/commits/d9848e5ddb39815d39b45f443b92a5c2a80e4fac .  Let me know what you think of that function I made, and specifically whether it should be exported from that location, and even more specifically whether we should be exporting it at all.

Reply:
I think I would just put it in the same test file (or at least in some `test/**/utils.ts` file or something, if for some reason you don't want to add it to that file). `src/types` is a public module, so doing unnecessary exports there is not a good idea.

Reply:
Addressed in https://github.com/nomiclabs/hardhat/pull/1928/commits/d57f8f1962ca22e6747861dab8456a0fb00e2da8